_source/_posts/2023-03-01-jakarta-ee-oidc.md

```diff
@@ -0,0 +1,546 @@
+                         FilterChain chain) throws IOException, ServletException {
+
+        HttpServletRequest request = (HttpServletRequest) servletRequest;
+        HttpServletResponse response = (HttpServletResponse) servletResponse;
+
+        LOGGER.info("In JwtFilter, path: " + request.getRequestURI());
+
+        // Get access token from authorization header
+        String authHeader = request.getHeader("authorization");
+        if (authHeader == null) {
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            response.getOutputStream().print("Unauthorized");
+            return;
+        } else {
+            String accessToken = authHeader.substring(authHeader.indexOf("Bearer ") + 7);
+            LOGGER.info("accesstoken: " + request.getRequestURI());
+            JwkProvider provider = new UrlJwkProvider(oidcConfig.getIssuerUri());
+            try {
+                DecodedJWT jwt = JWT.decode(accessToken);
+                // Get the kid from received JWT token
+                Jwk jwk = provider.get(jwt.getKeyId());
+
+                Algorithm algorithm = Algorithm.RSA256((RSAPublicKey) jwk.getPublicKey(), null);
+
+                JWTVerifier verifier = JWT.require(algorithm)
+                    .withIssuer(oidcConfig.getIssuerUri())
+                    .build();
+
+                jwt = verifier.verify(accessToken);
+                LOGGER.info("JWT decoded. sub=" + jwt.getClaims().get("sub"));
+                request.setAttribute("jwt", jwt);
+
+            } 
+            
+            ...
+
+        }
+
+        chain.doFilter(request, response);
+    }
+
+    @Override
+    public void destroy() {
+    }
+
+}
+```
+
+This code uses Auth0's JWT verifier for Java. Auth0 [has good docs on JWT verification](https://auth0.com/docs/secure/tokens/json-web-tokens/validate-json-web-tokens). If a valid JWT is found and decoded, it is saved in a request attribute, and the request is allowed to continue.
+
+Give it a try. Start the project.
+
+```bash
+./mvnw wildfly:run
+```
+
+Make a request to the protected API endpoint (not the OIDC endpoint).
+
+```bash
+http :8080/api/protected
+```
+
+You'll get:
+
+```bash
+HTTP/1.1 401 Unauthorized
+```
+
+Now, use your OIDC endpoint to retrieve a token. Using a browser, open `http://localhost:8080/protected`.
+
+Authenticate with Auth0. When you are redirected back to the protected servlet page, copy the token value and save it as a variable in a new shell.
+
+```bash
+TOKEN=eyJraWQiOiJqY3dpbGpUcGVZSG1Jajl6ODR3LV...
+```
+
+In that same shell, make a request to the protected API endpoint using the token.
+
+```bash
+http :8080/api/protected "Authorization: Bearer $TOKEN"
+```
+
+It should return something like the following.
+
+```bash
+HTTP/1.1 200 OK
+Connection: keep-alive
+Content-Length: 401
+Content-Type: text;charset=ISO-8859-1
+Date: Tue, 23 Feb 2023 07:48:59 GMT
+
+Welcome, andrew.hughes@mail.com
```

Comment:
@moksamedia This results in:

```
Welcome, "auth0|61bcbc76f64d4a0072af8a1d"
{sub="auth0|61bcbc76f64d4a0072af8a1d", aud=["https://dev-06bzs1cu.us.auth0.com/api/v2/","https://dev-06bzs1cu.us.auth0.com/userinfo"], azp="YwvfJ8soOxh3g1kmehMr1pXx6GgBIqMh", scope="openid profile email", iss="https://dev-06bzs1cu.us.auth0.com/", http://www.jakartaee.demo/roles=["Everyone","ROLE_ADMIN","ROLE_USER"], exp=1677268045, iat=1677181645}
```

Can you fix the code to display the user's email or update the post to reflect the proper `sub`?

Reply:
I updated to ApiFilter to pull the ID token from the userinfo endpoint and use that in the ApiServlet to show the email address.

Reply:
Thanks, good stuff!